File: ADSS Toolkit/autodrive_cpp/src/autodrive.cpp
// Import library
#include "autodrive.hpp"

////////////////////////////////////////////////////////////////////////////////

// VEHICLE CLASS METHODS

// Parse vehicle sensor data
void Vehicle::parse_data(json data, bool verbose=false) {
  // Actuator feedbacks
  throttle_str = data[1][id + " Throttle"];
  throttle = stof(throttle_str);
  steering_str = data[1][id + " Steering"];
  steering = stof(steering_str);
  // Wheel encoders
  encoder_ticks_str = data[1][id + " Encoder Ticks"];
  encoder_ticks = string_to_array(encoder_ticks_str);
  encoder_angles_str = data[1][id + " Encoder Angles"];
  encoder_angles = string_to_array(encoder_angles_str);
  // IPS
  position_str = data[1][id + " Position"];
  position = string_to_array(position_str);
  // IMU
  orientation_quaternion_str = data[1][id + " Orientation Quaternion"];
  orientation_quaternion = string_to_array(orientation_quaternion_str);
  orientation_euler_angles_str = data[1][id + " Orientation Euler Angles"];
  orientation_euler_angles = string_to_array(orientation_euler_angles_str);
  angular_velocity_str = data[1][id + " Angular Velocity"];
  angular_velocity = string_to_array(angular_velocity_str);
  linear_acceleration_str = data[1][id + " Linear Acceleration"];
  linear_acceleration = string_to_array(linear_acceleration_str);
  // LIDAR
  lidar_scan_rate_str = data[1][id + " LIDAR Scan Rate"];
  lidar_scan_rate = stof(lidar_scan_rate_str);
  lidar_range_array_str = data[1][id + " LIDAR Range Array"];
  lidar_range_array = string_to_array(lidar_range_array_str);
  lidar_intensity_array_str = data[1][id + " LIDAR Intensity Array"];
  lidar_intensity_array = string_to_array(lidar_intensity_array_str);
  // Cameras
  front_camera_image_str = data[1][id + " Front Camera Image"];
  front_camera_image = string_to_image(front_camera_image_str);
  rear_camera_image_str = data[1][id + " Rear Camera Image"];
  rear_camera_image = string_to_image(rear_camera_image_str);
  if (verbose) {
    cout << "\n--------------------------------" << endl;
    cout << "Receive Data from Vehicle: " << id << endl;
    cout << "--------------------------------" << endl;
    // Monitor vehicle data
    cout << "Throttle: " << throttle_str << endl;
    cout << "Steering: " << steering_str << endl;
    cout << "Encoder Ticks: " << encoder_ticks_str << endl;
    cout << "Encoder Angles: " << encoder_angles_str << endl;
    cout << "Position: " << position_str << endl;
    cout << "Orientation [Quaternion]: " << orientation_quaternion_str << endl;
    cout << "Orientation [Euler Angles]: " << orientation_euler_angles_str << endl;
    cout << "Angular Velocity: " << angular_velocity_str << endl;
    cout << "Linear Acceleration: " << linear_acceleration_str << endl;
    cout << "LIDAR Scan Rate: " << lidar_scan_rate_str << endl;
    cout << "LIDAR Range Array: " << endl << lidar_range_array_str << endl;
    cout << "LIDAR Intensity Array: " << endl << lidar_intensity_array_str << endl;
    resize(front_camera_image, front_camera_image, Size(640, 360));
    resize(rear_camera_image, rear_camera_image, Size(640, 360));
    imshow(id + " Front Camera Preview", front_camera_image);
    imshow(id + " Rear Camera Preview", rear_camera_image);
    waitKey(1);
  }
}

// Generate vehicle control commands
json Vehicle::generate_commands(json json_msg, bool verbose=false) {
  if (verbose) {
    cout << "\n-------------------------------" << endl;
    cout << "Transmit Data to Vehicle: " << id << endl;
    cout << "-------------------------------" << endl;
    // Monitor vehicle control commands
    throttle_command_str = to_string(throttle_command);
    steering_command_str = to_string(steering_command);
    cout << "Throttle Command: " << throttle_command_str << endl;
    cout << "Steering Command: " << steering_command_str << endl;
    if (headlights_command == 0) {
      headlights_command_str = "Disabled";
    }
    else if (headlights_command == 1) {
      headlights_command_str = "Low Beam";
    }
    else if (headlights_command == 2) {
      headlights_command_str = "High Beam";
    }
    else {
      headlights_command_str = "Invalid";
    }
    cout << "Headlights Command: " << headlights_command_str << endl;
    if (indicators_command == 0) {
      indicators_command_str = "Disabled";
    }
    else if (indicators_command == 1) {
      indicators_command_str = "Left Turn Indicator";
    }
    else if (indicators_command == 2) {
      indicators_command_str = "Right Turn Indicator";
    }
    else if (indicators_command == 3) {
      indicators_command_str = "Hazard Indicator";
    }
    else {
      indicators_command_str = "Invalid";
    }
    cout << "Indicators Command: " << indicators_command_str << endl;
  }
  json_msg[id + " Throttle"] = to_string(throttle_command); // Throttle command for vehicle
  json_msg[id + " Steering"] = to_string(steering_command); // Steering command for vehicle
  json_msg[id + " Headlights"] = to_string(headlights_command); // Headlights command for vehicle
  json_msg[id + " Indicators"] = to_string(indicators_command); // Indicators command for vehicle
  return json_msg;
}

////////////////////////////////////////////////////////////////////////////////

// TRAFFIC LIGHT CLASS METHODS

// Parse traffic light data
void TrafficLight::parse_data(json data, bool verbose=false) {
  // Traffic light state
  state_str = data[1][id + " State"];
  state = stoi(state_str);
  if (verbose) {
    cout << "\n--------------------------------------" << endl;
    cout << "Receive Data From Traffic Light: " << id << endl;
    cout << "--------------------------------------" << endl;
    // Monitor traffic light data
    if (state == 0) {
      state_str = "Disabled";
    }
    else if (state == 1) {
      state_str = "Red";
    }
    else if (state == 2) {
      state_str = "Yellow";
    }
    else if (state == 3) {
      state_str = "Green";
    }
    else {
      state_str = "Invalid";
    }
    cout << "Traffic Light State: " << state_str << endl;
  }
}

// Generate traffic light control commands
json TrafficLight::generate_commands(json json_msg, bool verbose=false) {
  if (verbose) {
    cout << "\n-------------------------------------" << endl;
    cout << "Transmit Data to Traffic Light: " << id << endl;
    cout << "-------------------------------------" << endl;
    // Monitor traffic light control commands
    if (command == 0) {
      command_str = "Disabled";
    }
    else if (command == 1) {
      command_str = "Red";
    }
    else if (command == 2) {
      command_str = "Yellow";
    }
    else if (command == 3) {
      command_str = "Green";
    }
    else {
      command_str = "Invalid";
    }
    cout << "Traffic Light Command: " << state_str << endl;
  }
  json_msg[id + " State"] = to_string(command); // Command for traffic light
  return json_msg;
}

File: ADSS Toolkit/autodrive_py/autodrive.py
#!/usr/bin/env python

# Import libraries
import numpy as np
import base64
from io import BytesIO
from PIL import Image
import cv2

################################################################################

# Nigel class
class Nigel:
    def __init__(self):
        # Nigel data
        self.id                       = None
        self.throttle                 = None
        self.steering                 = None
        self.encoder_ticks            = None
        self.encoder_angles           = None
        self.position                 = None
        self.orientation_quaternion   = None
        self.orientation_euler_angles = None
        self.angular_velocity         = None
        self.linear_acceleration      = None
        self.lidar_scan_rate          = None
        self.lidar_range_array        = None
        self.lidar_intensity_array    = None
        self.front_camera_image       = None
        self.rear_camera_image        = None
        # Nigel commands
        self.throttle_command   = None
        self.steering_command   = None
        self.headlights_command = None
        self.indicators_command = None

    # Parse Nigel sensor data
    def parse_data(self, data, verbose=False):
        # Actuator feedbacks
        self.throttle = float(data[self.id + " Throttle"])
        self.steering = float(data[self.id + " Steering"])
        # Wheel encoders
        self.encoder_ticks = np.fromstring(data[self.id + " Encoder Ticks"], dtype=int, sep=' ')
        self.encoder_angles = np.fromstring(data[self.id + " Encoder Angles"], dtype=float, sep=' ')
        # IPS
        self.position = np.fromstring(data[self.id + " Position"], dtype=float, sep=' ')
        # IMU
        self.orientation_quaternion = np.fromstring(data[self.id + " Orientation Quaternion"], dtype=float, sep=' ')
        self.orientation_euler_angles = np.fromstring(data[self.id + " Orientation Euler Angles"], dtype=float, sep=' ')
        self.angular_velocity = np.fromstring(data[self.id + " Angular Velocity"], dtype=float, sep=' ')
        self.linear_acceleration = np.fromstring(data[self.id + " Linear Acceleration"], dtype=float, sep=' ')
        # LIDAR
        self.lidar_scan_rate = float(data[self.id + " LIDAR Scan Rate"])
        self.lidar_range_array = np.fromstring(data[self.id + " LIDAR Range Array"], dtype=float, sep=' ')
        self.lidar_intensity_array = np.fromstring(data[self.id + " LIDAR Intensity Array"], dtype=float, sep=' ')
        # Cameras
        self.front_camera_image = cv2.cvtColor(np.asarray(Image.open(BytesIO(base64.b64decode(data[self.id + " Front Camera Image"])))), cv2.COLOR_RGB2BGR)
        self.rear_camera_image = cv2.cvtColor(np.asarray(Image.open(BytesIO(base64.b64decode(data[self.id + " Rear Camera Image"])))), cv2.COLOR_RGB2BGR)
        if verbose:
            print('\n--------------------------------')
            print('Receive Data from Nigel: ' + self.id)
            print('--------------------------------\n')
            # Monitor Nigel data
            print('Throttle: {}'.format(self.throttle))
            print('Steering: {}'.format(self.steering))
            print('Encoder Ticks:  {} {}'.format(self.encoder_ticks[0],self.encoder_ticks[1]))
            print('Encoder Angles: {} {}'.format(self.encoder_angles[0],self.encoder_angles[1]))
            print('Position: {} {} {}'.format(self.position[0],self.position[1],self.position[2]))
            print('Orientation [Quaternion]: {} {} {} {}'.format(self.orientation_quaternion[0],self.orientation_quaternion[1],self.orientation_quaternion[2],self.orientation_quaternion[3]))
            print('Orientation [Euler Angles]: {} {} {}'.format(self.orientation_euler_angles[0],self.orientation_euler_angles[1],self.orientation_euler_angles[2]))
            print('Angular Velocity: {} {} {}'.format(self.angular_velocity[0],self.angular_velocity[1],self.angular_velocity[2]))
            print('Linear Acceleration: {} {} {}'.format(self.linear_acceleration[0],self.linear_acceleration[1],self.linear_acceleration[2]))
            print('LIDAR Scan Rate: {}'.format(self.lidar_scan_rate))
            print('LIDAR Range Array: \n{}'.format(self.lidar_range_array))
            print('LIDAR Intensity Array: \n{}'.format(self.lidar_intensity_array))
            cv2.imshow(self.id + ' Front Camera Preview', cv2.resize(self.front_camera_image, (640, 360)))
            cv2.imshow(self.id + ' Rear Camera Preview', cv2.resize(self.rear_camera_image, (640, 360)))
            cv2.waitKey(1)

    # Generate Nigel control commands
    def generate_commands(self, verbose=False):
        if verbose:
            print('\n-------------------------------')
            print('Transmit Data to Nigel: ' + self.id)
            print('-------------------------------\n')
            # Monitor Nigel control commands
            print('Throttle Command: {}'.format(self.throttle_command))
            print('Steering Command: {}'.format(self.steering_command))
            if self.headlights_command == 0:
                headlights_cmd_str = 'Disabled'
            elif self.headlights_command == 1:
                headlights_cmd_str = 'Low Beam'
            elif self.headlights_command == 2:
                headlights_cmd_str = 'High Beam'
            else:
                headlights_cmd_str = 'Invalid'
            print('Headlights Command: {}'.format(headlights_cmd_str))
            if self.indicators_command == 0:
                indicators_cmd_str = 'Disabled'
            elif self.indicators_command == 1:
                indicators_cmd_str = 'Left Turn Indicator'
            elif self.indicators_command == 2:
                indicators_cmd_str = 'Right Turn Indicator'
            elif self.indicators_command == 3:
                indicators_cmd_str = 'Hazard Indicator'
            else:
                indicators_cmd_str = 'Invalid'
            print('Indicators Command: {}'.format(indicators_cmd_str))
        return {str(self.id) + ' Throttle': str(self.throttle_command), str(self.id) + ' Steering': str(self.steering_command), str(self.id) + ' Headlights': str(self.headlights_command), str(self.id) + ' Indicators': str(self.indicators_command)}

################################################################################

# F1TENTH class
class F1TENTH:
    def __init__(self):
        # F1TENTH data
        self.id                       = None
        self.throttle                 = None
        self.steering                 = None
        self.encoder_ticks            = None
        self.encoder_angles           = None
        self.position                 = None
        self.orientation_quaternion   = None
        self.orientation_euler_angles = None
        self.angular_velocity         = None
        self.linear_acceleration      = None
        self.lidar_scan_rate          = None
        self.lidar_range_array        = None
        self.lidar_intensity_array    = None
        self.front_camera_image       = None
        # F1TENTH commands
        self.throttle_command   = None
        self.steering_command   = None
    
    # Parse F1TENTH sensor data
    def parse_data(self, data, verbose=False):
        # Actuator feedbacks
        self.throttle = float(data[self.id + " Throttle"])
        self.steering = float(data[self.id + " Steering"])
        # Wheel encoders
        self.encoder_ticks = np.fromstring(data[self.id + " Encoder Ticks"], dtype=int, sep=' ')
        self.encoder_angles = np.fromstring(data[self.id + " Encoder Angles"], dtype=float, sep=' ')
        # IPS
        self.position = np.fromstring(data[self.id + " Position"], dtype=float, sep=' ')
        # IMU
        self.orientation_quaternion = np.fromstring(data[self.id + " Orientation Quaternion"], dtype=float, sep=' ')
        self.orientation_euler_angles = np.fromstring(data[self.id + " Orientation Euler Angles"], dtype=float, sep=' ')
        self.angular_velocity = np.fromstring(data[self.id + " Angular Velocity"], dtype=float, sep=' ')
        self.linear_acceleration = np.fromstring(data[self.id + " Linear Acceleration"], dtype=float, sep=' ')
        # LIDAR
        self.lidar_scan_rate = float(data[self.id + " LIDAR Scan Rate"])
        self.lidar_range_array = np.fromstring(data[self.id + " LIDAR Range Array"], dtype=float, sep=' ')
        self.lidar_intensity_array = np.fromstring(data[self.id + " LIDAR Intensity Array"], dtype=float, sep=' ')
        # Cameras
        self.front_camera_image = cv2.cvtColor(np.asarray(Image.open(BytesIO(base64.b64decode(data[self.id + " Front Camera Image"])))), cv2.COLOR_RGB2BGR)
        if verbose:
            print('\n--------------------------------')
            print('Receive Data from F1TENTH: ' + self.id)
            print('--------------------------------\n')
            # Monitor F1TENTH data
            print('Throttle: {}'.format(self.throttle))
            print('Steering: {}'.format(self.steering))
            print('Encoder Ticks:  {} {}'.format(self.encoder_ticks[0],self.encoder_ticks[1]))
            print('Encoder Angles: {} {}'.format(self.encoder_angles[0],self.encoder_angles[1]))
            print('Position: {} {} {}'.format(self.position[0],self.position[1],self.position[2]))
            print('Orientation [Quaternion]: {} {} {} {}'.format(self.orientation_quaternion[0],self.orientation_quaternion[1],self.orientation_quaternion[2],self.orientation_quaternion[3]))
            print('Orientation [Euler Angles]: {} {} {}'.format(self.orientation_euler_angles[0],self.orientation_euler_angles[1],self.orientation_euler_angles[2]))
            print('Angular Velocity: {} {} {}'.format(self.angular_velocity[0],self.angular_velocity[1],self.angular_velocity[2]))
            print('Linear Acceleration: {} {} {}'.format(self.linear_acceleration[0],self.linear_acceleration[1],self.linear_acceleration[2]))
            print('LIDAR Scan Rate: {}'.format(self.lidar_scan_rate))
            print('LIDAR Range Array: \n{}'.format(self.lidar_range_array))
            print('LIDAR Intensity Array: \n{}'.format(self.lidar_intensity_array))
            cv2.imshow(self.id + ' Front Camera Preview', cv2.resize(self.front_camera_image, (640, 360)))
            cv2.waitKey(1)

    # Generate F1TENTH control commands
    def generate_commands(self, verbose=False):
        if verbose:
            print('\n-------------------------------')
            print('Transmit Data to F1TENTH: ' + self.id)
            print('-------------------------------\n')
            # Monitor F1TENTH control commands
            print('Throttle Command: {}'.format(self.throttle_command))
            print('Steering Command: {}'.format(self.steering_command))
        return {str(self.id) + ' Throttle': str(self.throttle_command), str(self.id) + ' Steering': str(self.steering_command)}

################################################################################

# Traffic light class
class TrafficLight:
    def __init__(self):
        # Traffic light data
        self.id    = None
        self.state = None
        # Traffic light command
        self.command = None

    # Parse traffic light data
    def parse_data(self, data, verbose=False):
        # Traffic light state
        self.state = int(data[self.id + " State"])
        if verbose:
            print('\n--------------------------------------')
            print('Receive Data from Traffic Light: ' + self.id)
            print('--------------------------------------\n')
            # Monitor traffic light data
            if self.state == 0:
                state_str = 'Disabled'
            elif self.state == 1:
                state_str = 'Red'
            elif self.state == 2:
                state_str = 'Yellow'
            elif self.state == 3:
                state_str = 'Green'
            else:
                state_str = 'Invalid'
            print('Traffic Light State: {}'.format(state_str))

    # Generate traffic light control commands
    def generate_commands(self, verbose=False):
        if verbose:
            print('\n-------------------------------------')
            print('Transmit Data to Traffic Light: ' + self.id)
            print('-------------------------------------\n')
            # Monitor traffic light control commands
            if self.command == 0:
                command_str = 'Disabled'
            elif self.command == 1:
                command_str = 'Red'
            elif self.command == 2:
                command_str = 'Yellow'
            elif self.command == 3:
                command_str = 'Green'
            else:
                command_str = 'Invalid'
            print('Traffic Light Command: {}'.format(command_str))
        return {str(self.id) + ' State': str(self.command)}
